include/boost/gil/image_processing/threshold.hpp

```diff
@@ -0,0 +1,156 @@
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+//
+#ifndef BOOST_GIL_IMAGE_PROCESSING_THRESHOLD_HPP
+#define BOOST_GIL_IMAGE_PROCESSING_THRESHOLD_HPP
+
+#include <functional>
+#include <limits>
+#include <boost/gil/image.hpp>
+
+namespace boost { namespace gil {
+
+
+namespace detail {
+
+template
+<
+    typename SourceChannelT,
+    typename ResultChannelT,
+    typename SrcView,
+    typename DstView
+>
+void threshold_simple_impl
+(
+    SrcView const& src_view,
+    DstView const& dst_view,
+    std::function<ResultChannelT(SourceChannelT)> const& logic
+)
+{
+    //template argument validation
+    gil_function_requires<ImageViewConcept<SrcView>>();
+    gil_function_requires<MutableImageViewConcept<DstView>>();
+    gil_function_requires<ColorSpacesCompatibleConcept
+    <
+        typename color_space_type<SrcView>::type,
+        typename color_space_type<DstView>::type>
+    >();
+
+    //iterate over the image chaecking each pixel value for the threshold
+    for (size_t y = 0; y < src_view.height(); y++)
+    {
+        typename SrcView::x_iterator src_it = src_view.row_begin(y);
+        typename DstView::x_iterator dst_it = dst_view.row_begin(y);
+
+        for (size_t x = 0; x < src_view.width(); x++)
+        {
+            static_transform(src_it[x], dst_it[x], logic);
+        }
+    }
+}
+
+} //namespace boost::gil::detail
+
+enum class threshold_direction { regular, inverse };
+enum class threshold_optimal_value { otsu, triangle };
+enum class threshold_truncate_mode { threshold, zero };
+
+/*!
+Takes an image view and performes thresholding operation on each chennel.
+Values greater than or equal to threshold will be set to max
+If low is set to true then values more than thresholds are set to 0 else left unchanged.
+*/
+template <typename SrcView, typename DstView>
+void threshold_binary
+(
+    SrcView const& src_view,
+    DstView const& dst_view,
+    typename channel_type<DstView>::type threshold_value,
+    typename channel_type<DstView>::type max_value,
+    threshold_direction direction = threshold_direction::regular
+)
+{
+    //deciding output channel type and creating functor
+    typedef typename channel_type<SrcView>::type source_channel_t;
+    typedef typename channel_type<DstView>::type result_channel_t;
+
+    std::function<result_channel_t(source_channel_t)> threshold_logic =
+        (direction == threshold_direction::regular) ?
+        std::function<result_channel_t(source_channel_t)>(
+            [threshold_value, max_value](source_channel_t px) -> result_channel_t {
+                return px >= threshold_value ? max_value : 0; }) :
+        std::function<result_channel_t(source_channel_t)>(
+            [threshold_value, max_value](source_channel_t px) -> result_channel_t {
+                return px >= threshold_value ? 0 : max_value; });
+
+    detail::threshold_simple_impl<source_channel_t, result_channel_t>
```

Comment:
### Avoid std::function Part 3/3

I'd suggest to avoid use of `std::function` unless it is necessary.

I'd turn the use of ternary operator into this:

```
    if (direction == threshold_direction::regular)
    {
        detail::threshold_simple_impl(src_view, dst_view,
            [threshold_value, max_value](source_channel_t px) -> result_channel_t
                { return px >= threshold_value ? max_value : 0; });
    }
    else
    {
        detail::threshold_simple_impl(src_view, dst_view,
            [threshold_value, max_value](source_channel_t px) -> result_channel_t
                { return px >= threshold_value ? 0 : max_value; });
    }
```